test/testing_utils.py

```diff
@@ -0,0 +1,411 @@
+    optProb.finalize()
+    assertEqual(optProb.nObj, nObj)
+    assertEqual(optProb.nCon, nCon)
+    assertEqual(optProb.ndvs, nDV)
+
+
+def get_dict_distance(d, d2):
+    """This converts a flat 1D dict to array using sorted keys, then computes the L2 distance"""
+    assertEqual(set(d.keys()), set(d2.keys()))
+    a = []
+    a2 = []
+    for k in sorted(d.keys()):
+        a.append(d[k])
+        a2.append(d2[k])
+
+    a = np.array(a)
+    if a.ndim > 1:
+        a = np.concatenate(a)
+    a2 = np.array(a2)
+    if a2.ndim > 1:
+        a2 = np.concatenate(a2)
+    return np.linalg.norm(a - a2)
+
+
+# these are the informs if optimization completed successfully
+SUCCESS_INFORM = {
+    "SNOPT": 1,
+    "IPOPT": 0,
+    "NLPQLP": 0,
+    "SLSQP": 0,
+    "PSQP": 4,
+}
+# these are the name(s) of options that control the output file name
+OUTPUT_FILENAMES = {
+    "SNOPT": {"Print file": "_print.out", "Summary file": "_summary.out"},
+    "IPOPT": {"output_file": ".out"},
+    "SLSQP": {"IFILE": ".out"},
+    "PSQP": {"IFILE": ".out"},
+    "CONMIN": {"IFILE": ".out"},
+    "NLPQLP": {"iFile": ".out"},
+    "ParOpt": {"output_file": ".out"},
+    "ALPSO": {"filename": ".out"},
+    "NSGA2": {},
+}
+
+# these are optimizers which are installed by default
+DEFAULT_OPTIMIZERS = {"SLSQP", "PSQP", "CONMIN", "ALPSO", "NSGA2"}
+
+
+class OptTest(unittest.TestCase):
+    def assert_solution_allclose(self, sol, tol, partial_x=False):
+        """
+        An assertion method to check that the solution object matches the expected
+        optimum values defined in the class.
+
+        Parameters
+        ----------
+        sol : Solution object
+            The solution object after running an optimization
+        tol : float
+            The tolerance for assertions
+        partial_x : bool, optional
+            Whether partial assertion of the design variables ``x`` is allowed.
+            For large problems, we may not have the full x vector available at the optimum,
+            so we only check a few entries.
+        """
+        if not isinstance(self.xStar, list):
+            self.xStar = [self.xStar]
+        if not isinstance(self.fStar, list):
+            self.fStar = [self.fStar]
+        has_lambdaStar = hasattr(self, "lambdaStar") and hasattr(sol, "lambdaStar")
+        if has_lambdaStar and not isinstance(self.lambdaStar, list):
+            self.lambdaStar = [self.lambdaStar]
+
+        if not partial_x:
+            dist = []
+            for x in self.xStar:
+                dist.append(get_dict_distance(x, sol.xStar))
+            self.sol_index = dist.index(min(dist))
+        else:
+            # assume we have a single solution
+            self.sol_index = 0
+        # now we assert against the closest solution
+        # objective
+        # sol.fStar was broken for earlier versions of SNOPT
+        if self.optName == "SNOPT" and Version(self.optVersion) >= Version("7.7.7"):
```

Comment:
What does `Version` do here? It introduces an additional dependency on `packaging` and I'm not sure it's necessary.

Reply:
It allows for "correct" comparison of version strings, which is a non-trivial task (even though it's relatively simple here). `packaging` is a dependency of `setuptools` I believe, but I've switched to using `pkg_resources` which is something built into `setuptools` and therefore should available for everyone.

Reply:
Thanks